File: basic/basic_python/14888.py
import sys


# 연산 결과의 최대 또는 최소를 찾는 함수
# 한 연산에 대해 결과 저장 -> 그 결과를 가지고 함수 다시 호출

tmp_add = []
tmp_sub = []
tmp_mul = []
tmp_div = []


def set_tmp(mode):

    if len(tmp_add) >= 1 or len(tmp_sub) >= 1 or len(tmp_mul) >= 1 or len(tmp_div) >= 1:
        tmp_add.clear()
        tmp_sub.clear()
        tmp_mul.clear()
        tmp_div.clear()

    if mode == 'max':
        tmp_add.append(MIN_RESULT)
        tmp_sub.append(MIN_RESULT)
        tmp_mul.append(MIN_RESULT)
        tmp_div.append(MIN_RESULT)
    elif mode == 'min':
        tmp_add.append(MAX_RESULT)
        tmp_sub.append(MAX_RESULT)
        tmp_mul.append(MAX_RESULT)
        tmp_div.append(MAX_RESULT)

# current value의 초기값은 numbers[0]이 된다.
def find_max(current_value, operation_count, numbers, index):
    if index >= len(numbers):
        return

    isLast = True
    add_candidate, sub_candidate, mul_candidate, div_candidate = MIN_RESULT, MIN_RESULT, MIN_RESULT, MIN_RESULT

    if operation_count['add'] != 0:
        operation_count['add'] -= 1
        tmp_add[0] = current_value[0] + numbers[index + 1]
        find_max(tmp_add, operation_count, numbers, index + 1)
        add_candidate = max((add_candidate, tmp_add[0]))
        operation_count['add'] += 1
        isLast = False

    if operation_count['sub'] != 0:
        operation_count['sub'] -= 1
        tmp_sub[0] = current_value[0] + numbers[index + 1]
        find_max(tmp_sub, operation_count, numbers, index + 1)
        sub_candidate = max((sub_candidate, tmp_sub[0]))
        operation_count['sub'] += 1
        isLast = False

    if operation_count['mul'] != 0:
        operation_count['mul'] -= 1
        tmp_mul[0] = current_value[0] * numbers[index + 1]
        find_max(tmp_mul, operation_count, numbers, index + 1)
        mul_candidate = max((mul_candidate, tmp_mul[0]))
        operation_count['mul'] += 1
        isLast = False

    if operation_count['div'] != 0:
        operation_count['div'] -= 1
        if current_value[0] < 0:
            pos = -current_value[0]
            tmp_div[0] = pos // numbers[index + 1]
            tmp_div[0] *= -1
        else:
            tmp_div[0] = current_value[0] // numbers[index + 1]

        find_max(tmp_div, operation_count, numbers, index + 1)
        div_candidate = max((div_candidate, tmp_div[0]))
        operation_count['div'] += 1
        isLast = False

    if not isLast:
        current_value[0] = max((add_candidate, sub_candidate, mul_candidate, div_candidate))
    else:
        return


def find_min(current_value, operation_count, numbers, index):
    if index >= len(numbers):
        return

    isLast = True
    add_candidate, sub_candidate, mul_candidate, div_candidate = MAX_RESULT, MAX_RESULT, MAX_RESULT, MAX_RESULT

    if operation_count['add'] != 0:
        operation_count['add'] -= 1
        tmp_add[0] = current_value[0] + numbers[index + 1]
        find_min(tmp_add, operation_count, numbers, index + 1)
        add_candidate = min((add_candidate, tmp_add[0]))
        operation_count['add'] += 1
        isLast = False

    if operation_count['sub'] != 0:
        operation_count['sub'] -= 1
        tmp_sub[0] = current_value[0] - numbers[index + 1]
        find_min(tmp_sub, operation_count, numbers, index + 1)
        sub_candidate = min((sub_candidate, tmp_sub[0]))
        operation_count['sub'] += 1
        isLast = False

    if operation_count['mul'] != 0:
        operation_count['mul'] -= 1
        tmp_mul[0] = current_value[0] * numbers[index + 1]
        find_min(tmp_mul, operation_count, numbers, index + 1)
        mul_candidate = min((mul_candidate, tmp_mul[0]))
        operation_count['mul'] += 1
        isLast = False

    if operation_count['div'] != 0:
        operation_count['div'] -= 1
        if current_value[0] < 0:
            pos = -current_value[0]
            tmp_div[0] = pos // numbers[index + 1]
            tmp_div[0] *= -1
        else:
            tmp_div[0] = current_value[0] // numbers[index + 1]
        find_min(tmp_div, operation_count, numbers, index + 1)
        div_candidate = min((div_candidate, tmp_div[0]))
        operation_count['div'] += 1
        isLast = False

    if not isLast:
        current_value[0] = min((add_candidate, sub_candidate, mul_candidate, div_candidate))
    else:
        return

MIN_RESULT = -int(10e10)
MAX_RESULT = int(10e10)
num_count = int(sys.stdin.readline())
nums = list(map(int, sys.stdin.readline().split()))
operation = {}
operation['add'], operation['sub'], operation['mul'], operation['div'] = map(int, sys.stdin.readline().split())

max_val = []
min_val = []

max_val.append(nums[0])
min_val.append(nums[0])

find_max(max_val, operation, nums, 0)
find_min(min_val, operation, nums, 0)

print(max_val[0])
print(min_val[0])

File: basic/14888.cpp
#include <iostream>
#include <algorithm>

using namespace std;

enum
{
    ADD,
    SUB,
    MUL,
    DIV,
    MODE_NONE = -1,
    OP_NUM = 4,
    MAX_COUNT = 100
};

unsigned optr[OP_NUM];
int nums[MAX_COUNT];

const int MAX_VALUE = 10e10;
const int MIN_VALUE = -10e10;

int find_min(int current_value, size_t index, int);
int find_max(int current_value, size_t index, int);

int main(void)
{
    unsigned num_count;
    cin >> num_count;
    for(unsigned i = 0 ; i < num_count ; i++)
    {
        cin >> nums[i];
    }
    for(int i = 0 ; i < OP_NUM ; i++)
    {
        cin >> optr[i];
    }
    int max_val = find_max(nums[0], 0, MODE_NONE);
    int min_val = find_min(nums[0], 0, MODE_NONE);
    cout << max_val << '\n' << min_val;
    return 0;
}

int find_min(int current_value, size_t index, int mode)
{
    bool isLast = true;
    int tmp_add, tmp_sub, tmp_mul, tmp_div;
    tmp_add = MIN_VALUE;
    tmp_sub = MIN_VALUE;
    tmp_mul = MIN_VALUE;
    tmp_div = MIN_VALUE;

    if(optr[ADD] != 0)
    {
        tmp_add = current_value + nums[index + 1];
        optr[ADD] -= 1;
        tmp_add = find_min(tmp_add, index + 1, ADD);
        optr[ADD] += 1;
        isLast = false;
    }

    if(optr[SUB] != 0)
    {
        tmp_sub = current_value - nums[index + 1];
        optr[SUB] -= 1;
        tmp_sub = find_min(tmp_sub, index + 1, SUB);
        optr[SUB] += 1;
        isLast = false;
    }

    if(optr[MUL] != 0)
    {
        tmp_mul = current_value * nums[index + 1];
        optr[MUL] -= 1;
        tmp_mul = find_min(tmp_mul, index + 1, MUL);
        optr[MUL] += 1;
        isLast = false;
    }

    if(optr[DIV] != 0)
    {
        tmp_div = current_value / nums[index + 1];
        optr[DIV] -= 1;
        tmp_div = find_min(tmp_div, index + 1, DIV);
        optr[DIV] += 1;
        isLast = false;
    }

    if(!isLast)
    {
        current_value = min({tmp_add, tmp_sub, tmp_mul, tmp_div});
        return current_value;
    }
    else
    {
        if(mode == MODE_NONE)
        {
            return current_value;
        }

        if(mode == ADD)
        {
            return current_value + nums[index];
        }
        else if(mode == SUB)
        {
            return current_value - nums[index];
        }
        else if (mode == MUL)
        {
            return current_value * nums[index];
        }
        else
        {
            return current_value / nums[index];
        }
    }
}

int find_max(int current_value, size_t index, int mode)
{
    bool isLast = true;
    int tmp_add, tmp_sub, tmp_mul, tmp_div;
    tmp_add = MIN_VALUE;
    tmp_sub = MIN_VALUE;
    tmp_mul = MIN_VALUE;
    tmp_div = MIN_VALUE;

    if(optr[ADD] != 0)
    {
        tmp_add = current_value + nums[index + 1];
        optr[ADD] -= 1;
        tmp_add = find_max(tmp_add, index + 1, ADD);
        optr[ADD] += 1;
        isLast = false;
    }

    if(optr[SUB] != 0)
    {
        tmp_sub = current_value - nums[index + 1];
        optr[SUB] -= 1;
        tmp_sub = find_max(tmp_sub, index + 1, SUB);
        optr[SUB] += 1;
        isLast = false;
    }

    if(optr[MUL] != 0)
    {
        tmp_mul = current_value * nums[index + 1];
        optr[MUL] -= 1;
        tmp_mul = find_max(tmp_mul, index + 1, MUL);
        optr[SUB] += 1;
        isLast = false;
    }

    if(optr[DIV] != 0)
    {
        tmp_div = current_value / nums[index + 1];
        optr[DIV] -= 1;
        tmp_div = find_max(tmp_div, index + 1, DIV);
        optr[DIV] += 1;
        isLast = false;
    }
    
    if(!isLast)
    {
        current_value = max({tmp_add, tmp_sub, tmp_mul, tmp_div});
        return current_value;
    }
    else
    {
        if(mode == ADD)
        {
            return current_value + nums[index];
        }
        else if(mode == SUB)
        {
            return current_value - nums[index];
        }
        else if (mode == MUL)
        {
            return current_value * nums[index];
        }
        else
        {
            return current_value / nums[index];
        }
    }
}